### PR TITLE
fix: ETLng cleanup and graceful shutdown

### PR DESCRIPTION
**Pull request by godexsoft**
This PR implements graceful shutdown during initial ledger load. This is a bit of an edge case but good to have for completeness and robustness of the implementation.
Also in this PR some missing tests are added and general cleanup of the new ETL is performed.

### CODE REVIEW

File: src/etlng/ETLService.cpp [hunk 1]
```diff
@@ -223,13 +230,21 @@ ETLService::loadInitialLedgerIfNeeded()
...
-
-                    // TODO: this should be interruptible for graceful shutdown
-                    return loader_->loadInitialLedger(data);
-                });
+                return extractor_->extractLedgerOnly(seq).and_then(
+                    [this, seq](auto&& data) -> std::optional<ripple::LedgerHeader> {
+                        // TODO: loadInitialLedger in balancer should be called fetchEdgeKeys or similar
+                        auto res = balancer_->loadInitialLedger(seq, *initialLoadObserver_);
+                        if (not res.has_value() and res.error() == InitialLedgerLoadError::Cancel) {
+                            LOG(log_.info()) << "Initial ledger load got cancelled";
```
Comment (mathbunnyru) [inline src/etlng/ETLService.cpp]: Maybe warn here as well?

File: src/etlng/LoadBalancer.cpp [hunk 2]
```diff
@@ -210,30 +210,34 @@ LoadBalancer::LoadBalancer(
...
-
-            if (!res) {
-                LOG(log_.error()) << "Failed to download initial ledger."
-                                  << " Sequence = " << sequence << " source = " << source->toString();
-            } else {
-                response = std::move(data);
+            auto res = source->loadInitialLedger(sequence, downloadRanges_, loadObserver);
+
+            if (not res.has_value()) {
+                if (res.error() == InitialLedgerLoadError::Error) {
```
Comment (mathbunnyru) [inline src/etlng/LoadBalancer.cpp]: Let's do a switch without `default` here, because we might add new cases, and compiler will tell us we forgot about them

Comment (mathbunnyru) [reply src/etlng/LoadBalancer.cpp]: and consider it for other places

File: src/etlng/LoadBalancerInterface.hpp [hunk 3]
```diff
@@ -39,6 +39,20 @@
 
 namespace etlng {
 
+/**
+ * @brief Represents possible errors for initial ledger load
+ */
+enum class InitialLedgerLoadError {
+    Cancel, /*< Indicating the initial load got cancelled by user */
+    Error,  /*< Indicating some error happened during initial ledger load */
```
Comment (mathbunnyru) [inline src/etlng/LoadBalancerInterface.hpp]: Maybe we could put comments as part of doxygen?
Or will they be generated already?

File: src/etlng/LoadBalancerInterface.hpp [hunk 4]
```diff
@@ -39,6 +39,20 @@
...
+};
+
+/**
+ * @brief The result type of the initial ledger load
+ * @note The value channel represents edge keys
```
Comment (mathbunnyru) [inline src/etlng/LoadBalancerInterface.hpp]: channel?

File: src/etlng/LoaderInterface.hpp [hunk 5]
```diff
@@ -25,11 +25,16 @@
...
 #include <optional>
-#include <string>
 
 namespace etlng {
 
-using Error = std::string;
+/**
+ * @brief Enumeration of possible errors that can occur during loading operations
+ */
+enum class LoaderError {
```
Comment (mathbunnyru) [inline src/etlng/LoaderInterface.hpp:34]: Nice, that's what I was talking about int the previous PR 🙂

File: src/etlng/impl/TaskManager.cpp [hunk 6]
```diff
@@ -121,15 +121,21 @@ TaskManager::spawnLoader(TaskQueue& queue)
...
+                        LOG(log_.warn()) << "Immediately stopping loader on write conflict"
+                                         << "; latest ledger cache loaded for " << data->seq;
+                        monitor_.get().notifyWriteConflict(data->seq);
+                        break;
+                    }
+
+                    if (expectedSuccess.error() == LoaderError::AmendmentBlocked) {
+                        LOG(log_.warn()) << "Immediately stopping loader on amendment block";
+                        break;
+                    }
```
Comment (mathbunnyru) [inline src/etlng/impl/TaskManager.cpp]: definitely switch here

**Review by mathbunnyru — CHANGES_REQUESTED** [review]
Overall, looks good

File: src/etlng/ETLService.cpp (hunk 1 above)
Comment (godexsoft) [reply src/etlng/ETLService.cpp]: I was considering to actually lower this log to debug not to pollute the log with similar messages

File: src/etlng/LoadBalancer.cpp (hunk 2 above)
Comment (godexsoft) [reply src/etlng/LoadBalancer.cpp]: But there is only one interesting case here. I don't see the benefit in this case.
In the other one switch aint gonna work because we use `break` there, hence no switch

File: src/etlng/LoadBalancerInterface.hpp (hunk 3 above)
Comment (godexsoft) [reply src/etlng/LoadBalancerInterface.hpp]: These are doxygen comments.

File: src/etlng/LoadBalancerInterface.hpp (hunk 4 above)
Comment (godexsoft) [reply src/etlng/LoadBalancerInterface.hpp]: Expected got value and error, they are commonly called "channels".

File: src/etlng/LoaderInterface.hpp (hunk 5 above)
Comment (godexsoft) [reply src/etlng/LoaderInterface.hpp:34]: It made sense to change from string because of the difference in how we treat these two 👍

File: src/etlng/LoadBalancerInterface.hpp (hunk 3 above)
Comment (mathbunnyru) [reply src/etlng/LoadBalancerInterface.hpp]: Cool 👍

File: src/etlng/impl/TaskManager.cpp (hunk 6 above)
Comment (godexsoft) [reply src/etlng/impl/TaskManager.cpp]: Feel free to propose how to change it so it's a switch 😃

File: src/etlng/ETLService.cpp (hunk 1 above)
Comment (mathbunnyru) [reply src/etlng/ETLService.cpp]: I think it's reasonable

File: src/etlng/impl/TaskManager.cpp (hunk 6 above)
Comment (mathbunnyru) [reply src/etlng/impl/TaskManager.cpp]: Yes, you're right, no good way right now

File: src/etlng/ETLService.cpp [hunk 7]
```diff
@@ -223,13 +230,21 @@ ETLService::loadInitialLedgerIfNeeded()
...
-                    data.edgeKeys = balancer_->loadInitialLedger(seq, *initialLoadObserver_);
-
-                    // TODO: this should be interruptible for graceful shutdown
-                    return loader_->loadInitialLedger(data);
-                });
+                return extractor_->extractLedgerOnly(seq).and_then(
+                    [this, seq](auto&& data) -> std::optional<ripple::LedgerHeader> {
+                        // TODO: loadInitialLedger in balancer should be called fetchEdgeKeys or similar
+                        auto res = balancer_->loadInitialLedger(seq, *initialLoadObserver_);
+                        if (not res.has_value() and res.error() == InitialLedgerLoadError::Cancel) {
```
Comment (kuznetsss) [inline src/etlng/ETLService.cpp]: Shouldn't switch-case here be used to make sure we have all the possible errors covered?

Also maybe error name could be `Cancelled` or `OperationAborted`?

File: src/etlng/LoadBalancer.cpp (hunk 2 above)
Comment (mathbunnyru) [reply src/etlng/LoadBalancer.cpp]: Got it, I think you can combine 2 if statements in 1 here though

File: src/etlng/LoadBalancerInterface.hpp (hunk 4 above)
Comment (kuznetsss) [reply src/etlng/LoadBalancerInterface.hpp]: Channel is confusing for me as well. Usually channel is a thread safe queue.

File: src/etlng/impl/GrpcSource.cpp [hunk 8]
```diff
@@ -131,9 +136,9 @@ GrpcSource::loadInitialLedger(
         ASSERT(tag != nullptr, "Tag can't be null.");
         auto ptr = static_cast<AsyncGrpcCall*>(tag);
 
-        if (!ok) {
-            LOG(log_.error()) << "loadInitialLedger - ok is false";
-            return {{}, false};  // cancelled
+        if (not ok or initialLoadShouldStop_->load()) {
```
Comment (kuznetsss) [inline src/etlng/impl/GrpcSource.cpp]: Shouldn't we check the flag at the beginning of the method as well if somehow `loadInitialLedger()` would be called after `stop()`? Or is it not possible?

File: src/etlng/impl/TaskManager.cpp (hunk 6 above)
Comment (kuznetsss) [reply src/etlng/impl/TaskManager.cpp]: I agree with @mathbunnyru, if you are using an `enum`, it should be processed with a `switch`. It will save us from forgetting to cover new enum's variants in the future.

File: src/etlng/ETLService.cpp (hunk 7 above)
Comment (godexsoft) [reply src/etlng/ETLService.cpp]: We just had the same convo with @mathbunnyru - we are only interested in one case here so it does not make sense to use switch imo. (if the logic changes we can always adjust).

File: src/etlng/LoadBalancerInterface.hpp (hunk 4 above)
Comment (godexsoft) [reply src/etlng/LoadBalancerInterface.hpp]: This is not golang 😃

Comment (godexsoft) [reply src/etlng/LoadBalancerInterface.hpp]: On a more serious note, what would you call it so i can adjust? @kuznetsss @mathbunnyru

File: src/etlng/impl/GrpcSource.cpp (hunk 8 above)
Comment (godexsoft) [reply src/etlng/impl/GrpcSource.cpp]: No harm in adding it next to the stub check i guess

File: src/etlng/LoadBalancerInterface.hpp (hunk 4 above)
Comment (mathbunnyru) [reply src/etlng/LoadBalancerInterface.hpp]: The successful value represents edge keys

File: src/etlng/impl/TaskManager.cpp (hunk 6 above)
Comment (godexsoft) [reply src/etlng/impl/TaskManager.cpp]: @kuznetsss i challenge you to implement it then 😃

Comment (godexsoft) [reply src/etlng/impl/TaskManager.cpp]: Fine, let's use immediately invoked lambda with a switch inside. not super clean but addresses your concerns

File: tests/unit/etlng/ETLServiceTests.cpp [hunk 9]
```diff
@@ -488,9 +489,7 @@ TEST_F(ETLServiceTests, GiveUpWriterAfterWriteConflict)
     EXPECT_FALSE(systemState_->writeConflict);  // and removes write conflict flag
 }
 
-struct ETLServiceAssertTests : common::util::WithMockAssert, ETLServiceTests {};
```
Comment (mathbunnyru) [inline tests/unit/etlng/ETLServiceTests.cpp:491]: Why did you remove this?

File: src/etlng/ETLService.cpp [hunk 10]
```diff
@@ -136,7 +139,11 @@ ETLService::run()
             return;
         }
 
-        ASSERT(rng.has_value(), "Ledger range can't be null");
+        if (not rng.has_value()) {
+            LOG(log_.warn()) << "Initial ledger download got cancelled - stopping ETL service";
```
Comment (mathbunnyru) [inline src/etlng/ETLService.cpp:143]: Don't we have to check the error here?

File: src/etlng/impl/GrpcSource.cpp [hunk 11]
```diff
@@ -131,9 +140,9 @@ GrpcSource::loadInitialLedger(
         ASSERT(tag != nullptr, "Tag can't be null.");
         auto ptr = static_cast<AsyncGrpcCall*>(tag);
 
-        if (!ok) {
-            LOG(log_.error()) << "loadInitialLedger - ok is false";
-            return {{}, false};  // cancelled
+        if (not ok or *initialLoadShouldStop_) {
```
Comment (mathbunnyru) [inline src/etlng/impl/GrpcSource.cpp:143]: I don't like having `ok` var here, because it looks like we're just hiding errors, maybe there is a better name?

File: tests/unit/etlng/ETLServiceTests.cpp [hunk 12]
```diff
@@ -501,10 +500,10 @@ TEST_F(ETLServiceAssertTests, FailToLoadInitialLedger)
     EXPECT_CALL(*loader_, loadInitialLedger).Times(0);
     EXPECT_CALL(*taskManagerProvider_, make).Times(0);
 
-    EXPECT_CLIO_ASSERT_FAIL({ service_.run(); });
```
Comment (mathbunnyru) [inline tests/unit/etlng/ETLServiceTests.cpp:504]: Why this change?

**Review by mathbunnyru — APPROVED** [review]
Left some questions - if you think these places are ok, don't want to prevent you from merging, so 👍

File: tests/unit/etlng/ETLServiceTests.cpp (hunk 9 above)
Comment (godexsoft) [reply tests/unit/etlng/ETLServiceTests.cpp:491]: Because these asserts are gone too now that we can potentially stop the loading of initial ledger 👍

File: src/etlng/ETLService.cpp (hunk 10 above)
Comment (godexsoft) [reply src/etlng/ETLService.cpp:143]: The only reason to get here is cancellation - there is an assert below checking for invalid logic. Any error inside should trigger retry until success or the end of the universe, whichever comes first

File: tests/unit/etlng/ETLServiceTests.cpp (hunk 12 above)
Comment (godexsoft) [reply tests/unit/etlng/ETLServiceTests.cpp:504]: Assert is gone

File: src/etlng/impl/GrpcSource.cpp (hunk 11 above)
Comment (godexsoft) [reply src/etlng/impl/GrpcSource.cpp:143]: I don't like the entire grpc way of async downloading and frankly idk what this 'ok' is - it copied over from the old logic and i don't know enough about it to confidently change it atm
